File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/recommended_fee/RecommendedFeeFragment.java
package io.muun.apollo.presentation.ui.fragments.recommended_fee;


import io.muun.apollo.R;
import io.muun.apollo.domain.model.CurrencyDisplayMode;
import io.muun.apollo.domain.model.FeeOption;
import io.muun.apollo.domain.model.PaymentAnalysis;
import io.muun.apollo.domain.model.PaymentContext;
import io.muun.apollo.domain.model.PaymentRequest;
import io.muun.apollo.presentation.ui.base.SingleFragment;
import io.muun.apollo.presentation.ui.new_operation.TitleAndDescriptionDrawer;
import io.muun.apollo.presentation.ui.view.FeeManualItem;
import io.muun.apollo.presentation.ui.view.FeeOptionItem;
import io.muun.apollo.presentation.ui.view.HtmlTextView;
import io.muun.apollo.presentation.ui.view.MuunButton;
import io.muun.apollo.presentation.ui.view.MuunHeader.Navigation;
import io.muun.apollo.presentation.ui.view.RichText;
import io.muun.apollo.presentation.ui.view.StatusMessage;
import io.muun.common.Rules;
import io.muun.common.utils.Preconditions;

import android.text.TextUtils;
import android.view.View;
import butterknife.BindString;
import butterknife.BindView;
import butterknife.OnClick;
import icepick.State;

public class RecommendedFeeFragment extends SingleFragment<RecommendedFeePresenter>
        implements RecommendedFeeView {

    @BindView(R.id.fee_options_message)
    HtmlTextView message;

    @BindView(R.id.fee_option_fast)
    FeeOptionItem feeOptionItemFast;

    @BindView(R.id.fee_option_medium)
    FeeOptionItem feeOptionItemMedium;

    @BindView(R.id.fee_option_slow)
    FeeOptionItem feeOptionItemSlow;

    @BindView(R.id.enter_fee_manually)
    FeeManualItem feeManualItem;

    @BindView(R.id.status_message)
    StatusMessage statusMessage;

    @BindView(R.id.confirm_fee)
    MuunButton confirmButton;

    @BindString(R.string.fee_options_message)
    String messageText;

    @BindString(R.string.fee_options_whats_this)
    String whatsThisText;

    private Double selectedFeeRate;

    @State
    CurrencyDisplayMode currencyDisplayMode;

    @Override
    protected void inject() {
        getComponent().inject(this);
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.recommended_fee_selection_fragment;
    }

    @Override
    protected void initializeUi(View view) {
        super.initializeUi(view);

        getParentActivity().getHeader().setNavigation(Navigation.EXIT);

        final CharSequence content = TextUtils.concat(
                messageText,
                ". ",
                new RichText(whatsThisText).setLink(this::onWhatsThisClick)
        );

        message.setText(content);

        if (selectedFeeRate == null) {
            confirmButton.setEnabled(false);
        }
    }

    @Override
    public void setCurrencyDisplayMode(CurrencyDisplayMode currencyDisplayMode) {
        this.currencyDisplayMode = currencyDisplayMode;
    }

    @Override
    public void setPaymentContext(PaymentContext payCtx, PaymentRequest payReq) {

        final FeeOption feeOptionFast = payCtx.getFastFeeOption();
        final FeeOption feeOptionMid = payCtx.getMediumFeeOption();
        final FeeOption feeOptionSlow = payCtx.getSlowFeeOption();

        analyzeAndBindFeeOption(feeOptionItemFast, payCtx, payReq, feeOptionFast);
        analyzeAndBindFeeOption(feeOptionItemMedium, payCtx, payReq, feeOptionMid);
        analyzeAndBindFeeOption(feeOptionItemSlow, payCtx, payReq, feeOptionSlow);

        if (payReq.getTakeFeeFromAmount()) {
            statusMessage.setWarning(
                    R.string.use_all_funds_warning_message,
                    R.string.fee_options_use_all_funds_warning_desc,
                    false,
                    ':'
            );
        }

        final double feeRateFast = feeOptionFast.getSatoshisPerByte();
        final double feeRateMid = feeOptionMid.getSatoshisPerByte();
        final double feeRateSlow = feeOptionSlow.getSatoshisPerByte();
        hideDuplicatedFeeRateOptions(feeRateFast, feeRateMid, feeRateSlow);

        final double currentFeeRate = Preconditions.checkNotNull(payReq.getFeeInSatoshisPerByte());

        final PaymentAnalysis analysis;
        try {
            analysis = payCtx.analyze(payReq);

        } catch (Throwable error) {
            presenter.handleError(error);
            return;
        }

        if (alreadySomeOptionSelected()) { // then we probably resuming from background, already set
            return;
        }

        confirmButton.setEnabled(false); // will be enabled later if selected fee is OK

        if (analysis.getCanPayWithSelectedFee()) {
            if (Rules.feeRateEquals(currentFeeRate, feeRateFast)) {
                selectFeeOption(feeOptionFast, feeOptionItemFast);

            } else if (Rules.feeRateEquals(currentFeeRate, feeRateMid)) {
                selectFeeOption(feeOptionMid, feeOptionItemMedium);

            } else if (Rules.feeRateEquals(currentFeeRate, feeRateSlow)) {
                selectFeeOption(feeOptionSlow, feeOptionItemSlow);

            } else {
                showManuallySelectedFee(analysis);
            }
        }
    }

    private boolean alreadySomeOptionSelected() {
        return feeOptionItemFast.isSelected()
                || feeOptionItemMedium.isSelected()
                || feeOptionItemSlow.isSelected()
                || feeManualItem.isSelected();
    }

    private void showManuallySelectedFee(PaymentAnalysis analysis) {
        Preconditions.checkNotNull(analysis);

        if (analysis.getCanPayWithSelectedFee()) {
            confirmButton.setEnabled(true);
            selectedFeeRate = analysis.getPayReq().getFeeInSatoshisPerByte();
        }

        feeManualItem.setCurrencyDisplayMode(currencyDisplayMode);
        feeManualItem.setAmount(analysis.getFee().inInputCurrency);
        feeManualItem.setSelected(true);
    }

    private void analyzeAndBindFeeOption(FeeOptionItem feeOptionItem,
                                         PaymentContext paymentContext,
                                         PaymentRequest payReq,
                                         FeeOption feeOption) {

        final PaymentAnalysis analysis;
        try {
            analysis = paymentContext.analyze(payReq.withFeeRate(feeOption.getSatoshisPerByte()));

        } catch (Throwable error) {
            presenter.handleError(error);
            return;
        }

        bindFeeOptionItem(feeOptionItem, analysis, feeOption);
    }

    private void bindFeeOptionItem(FeeOptionItem feeOptionItem,
                                   PaymentAnalysis analysis,
                                   FeeOption feeOption) {

        feeOptionItem.setCurrencyDisplayMode(currencyDisplayMode);
        feeOptionItem.setMaxTimeMs(feeOption.getMaxTimeMs());
        feeOptionItem.setFeeRate(feeOption.getSatoshisPerByte());
        feeOptionItem.setFee(analysis.getFee());

        if (analysis.getCanPayWithSelectedFee()) {

            feeOptionItem.setOnClickListener(v -> {
                selectFeeOption(feeOption, feeOptionItem);
            });

        } else {
            feeOptionItem.setEnabled(false);
        }
    }

    private void hideDuplicatedFeeRateOptions(double fastRate, double mediumRate, double slowRate) {
        if (Rules.feeRateEquals(mediumRate, slowRate)) {
            feeOptionItemSlow.setVisibility(View.GONE);
        }

        if (Rules.feeRateEquals(fastRate, mediumRate)) {
            feeOptionItemMedium.setVisibility(View.GONE);
        }
    }

    private void selectFeeOption(FeeOption option, FeeOptionItem item) {
        selectedFeeRate = option.getSatoshisPerByte();

        feeOptionItemFast.setSelected(false);
        feeOptionItemMedium.setSelected(false);
        feeOptionItemSlow.setSelected(false);
        feeManualItem.setSelected(false);

        item.setSelected(true);
        confirmButton.setEnabled(true);
    }

    private void onWhatsThisClick() {
        final TitleAndDescriptionDrawer dialog = new TitleAndDescriptionDrawer();
        dialog.setTitle(R.string.fee_options_whats_this_explanation_title);
        dialog.setDescription(getString(R.string.fee_options_whats_this_explanation_desc));
        showDrawerDialog(dialog);

        presenter.reportShowSelectFeeInfo();
    }

    @OnClick(R.id.enter_fee_manually)
    void onEditFeeManuallyClick() {
        presenter.editFeeManually();
    }

    @OnClick(R.id.confirm_fee)
    void onConfirmFeeClick() {
        presenter.confirmFee(selectedFeeRate);
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/fragments/recommended_fee/RecommendedFeeFragment.kt
package io.muun.apollo.presentation.ui.fragments.recommended_fee

import android.text.TextUtils
import android.view.View
import butterknife.BindString
import butterknife.BindView
import icepick.State
import io.muun.apollo.R
import io.muun.apollo.domain.model.BitcoinAmount
import io.muun.apollo.domain.model.BitcoinUnit
import io.muun.apollo.presentation.ui.base.SingleFragment
import io.muun.apollo.presentation.ui.new_operation.TitleAndDescriptionDrawer
import io.muun.apollo.presentation.ui.new_operation.estimateTimeInMs
import io.muun.apollo.presentation.ui.view.FeeManualItem
import io.muun.apollo.presentation.ui.view.FeeOptionItem
import io.muun.apollo.presentation.ui.view.HtmlTextView
import io.muun.apollo.presentation.ui.view.MuunButton
import io.muun.apollo.presentation.ui.view.MuunHeader
import io.muun.apollo.presentation.ui.view.RichText
import io.muun.apollo.presentation.ui.view.StatusMessage
import io.muun.common.Rules
import newop.EditFeeState
import newop.FeeState

class RecommendedFeeFragment : SingleFragment<RecommendedFeePresenter>(), RecommendedFeeView {

    @BindView(R.id.fee_options_message)
    lateinit var message: HtmlTextView

    @BindView(R.id.fee_option_fast)
    lateinit var feeOptionItemFast: FeeOptionItem

    @BindView(R.id.fee_option_medium)
    lateinit var feeOptionItemMedium: FeeOptionItem

    @BindView(R.id.fee_option_slow)
    lateinit var feeOptionItemSlow: FeeOptionItem

    @BindView(R.id.enter_fee_manually)
    lateinit var feeManualItem: FeeManualItem

    @BindView(R.id.status_message)
    lateinit var statusMessage: StatusMessage

    @BindView(R.id.confirm_fee)
    lateinit var confirmButton: MuunButton

    @BindString(R.string.fee_options_message)
    lateinit var messageText: String

    @BindString(R.string.fee_options_whats_this)
    lateinit var whatsThisText: String

    private var selectedFeeRateInVBytes: Double? = null // Starts with null if no rate zpre-selected

    @State
    lateinit var mBitcoinUnit: BitcoinUnit

    override fun inject() {
        component.inject(this)
    }

    override fun getLayoutResource(): Int {
        return R.layout.recommended_fee_selection_fragment
    }

    override fun initializeUi(view: View) {
        message.text = TextUtils.concat(
            messageText,
            ". ",
            RichText(whatsThisText).setLink { onWhatsThisClick() }
        )

        if (selectedFeeRateInVBytes == null) {
            confirmButton.isEnabled = false
        }

        feeManualItem.setOnClickListener {
            presenter!!.editFeeManually()
        }

        confirmButton.setOnClickListener {
            presenter!!.confirmFee(selectedFeeRateInVBytes!!) // NonNull (feeRate has been selected)
        }
    }

    override fun setUpHeader() {
        parentActivity.header.setNavigation(MuunHeader.Navigation.EXIT)
        parentActivity.header.showTitle(R.string.edit_fee_title)
    }

    override fun setBitcoinUnit(bitcoinUnit: BitcoinUnit) {
        this.mBitcoinUnit = bitcoinUnit
    }

    override fun setState(state: EditFeeState) {

        val feeWindow = state.resolved.paymentContext.feeWindow
        bindFeeOption(feeOptionItemFast, state, feeWindow.fastConfTarget)
        bindFeeOption(feeOptionItemMedium, state, feeWindow.mediumConfTarget)
        bindFeeOption(feeOptionItemSlow, state, feeWindow.slowConfTarget)

        if (state.amountInfo.takeFeeFromAmount) {
            statusMessage.setWarning(
                R.string.use_all_funds_warning_message,
                R.string.fee_options_use_all_funds_warning_desc,
                false,
                ':'
            )
        }

        val feeRateFastInVBytes = state.minFeeRateForTarget(feeWindow.fastConfTarget)
        val feeRateMidInVBytes = state.minFeeRateForTarget(feeWindow.mediumConfTarget)
        val feeRateSlowInVBytes = state.minFeeRateForTarget(feeWindow.slowConfTarget)

        hideDuplicatedFeeRateOptions(feeRateFastInVBytes, feeRateMidInVBytes, feeRateSlowInVBytes)

        val currentFeeRateInVBytes = state.amountInfo.feeRateInSatsPerVByte
        val feeData = state.calculateFee(currentFeeRateInVBytes)

        if (alreadySomeOptionSelected()) { // then we probably resuming from background, already set
            return
        }

        confirmButton.isEnabled = false // will be enabled later if selected fee is OK

        if (feeData.isFinal) {
            when {
                Rules.feeRateEquals(currentFeeRateInVBytes, feeRateFastInVBytes) ->
                    selectFeeOption(feeOptionItemFast, feeRateFastInVBytes)

                Rules.feeRateEquals(currentFeeRateInVBytes, feeRateMidInVBytes) ->
                    selectFeeOption(feeOptionItemMedium, feeRateMidInVBytes)

                Rules.feeRateEquals(currentFeeRateInVBytes, feeRateSlowInVBytes) ->
                    selectFeeOption(feeOptionItemSlow, feeRateSlowInVBytes)

                else ->
                    showManuallySelectedFee(feeData, currentFeeRateInVBytes)
            }
        }
    }

    private fun alreadySomeOptionSelected(): Boolean =
        feeOptionItemFast.isSelected
            || feeOptionItemMedium.isSelected
            || feeOptionItemSlow.isSelected
            || feeManualItem.isSelected

    private fun showManuallySelectedFee(feeData: FeeState, currentFeeRateInVBytes: Double) {
        if (feeData.isFinal) {
            confirmButton.isEnabled = true
            selectedFeeRateInVBytes = currentFeeRateInVBytes
        }
        feeManualItem.bitcoinUnit = mBitcoinUnit
        feeManualItem.amount = BitcoinAmount.fromLibwallet(feeData.amount).inInputCurrency
        feeManualItem.isSelected = true
    }

    private fun bindFeeOption(feeOptionItem: FeeOptionItem, state: EditFeeState, confTarget: Long) {

        val feeRateInVBytes = state.minFeeRateForTarget(confTarget)

        // TODO currently forgets input currency, which is important for amount display
        val feeData = state.calculateFee(feeRateInVBytes)

        feeOptionItem.setBitcoinUnit(mBitcoinUnit)
        feeOptionItem.setMaxTimeMs(estimateTimeInMs(feeData.targetBlocks.toInt()))
        feeOptionItem.setFeeRate(feeRateInVBytes)
        feeOptionItem.setFee(BitcoinAmount.fromLibwallet(feeData.amount))

        if (feeData.isFinal) {
            feeOptionItem.setOnClickListener {
                selectFeeOption(feeOptionItem, feeRateInVBytes)
            }
        } else {
            feeOptionItem.isEnabled = false
        }
    }

    private fun hideDuplicatedFeeRateOptions(fast: Double, medium: Double, slow: Double) {

        if (Rules.feeRateEquals(medium, slow)) {
            feeOptionItemSlow.visibility = View.GONE
        }

        if (Rules.feeRateEquals(fast, medium)) {
            feeOptionItemMedium.visibility = View.GONE
        }
    }

    private fun selectFeeOption(item: FeeOptionItem, feeRateInVBytes: Double) {
        selectedFeeRateInVBytes = feeRateInVBytes

        feeOptionItemFast.isSelected = false
        feeOptionItemMedium.isSelected = false
        feeOptionItemSlow.isSelected = false
        feeManualItem.isSelected = false

        item.isSelected = true
        confirmButton.isEnabled = true
    }

    private fun onWhatsThisClick() {
        val dialog = TitleAndDescriptionDrawer()
        dialog.setTitle(R.string.fee_options_whats_this_explanation_title)
        dialog.setDescription(getString(R.string.fee_options_whats_this_explanation_desc))
        showDrawerDialog(dialog)
        presenter!!.reportShowSelectFeeInfo()
    }
}